cpp/dolfinx/fem/DirichletBC.h

```diff
@@ -495,76 +508,71 @@ class DirichletBC
   void set(std::span<T> x, std::optional<std::span<const T>> x0,
            T alpha = 1) const
   {
-    std::int32_t x_size = x.size();
-    if (alpha == T(0)) // Optimisation for when alpha == 0
+    // set_fn is a lambda which gets evaluated for every index in [0,
+    // _dofs0.size()) and its result is assigned to x[_dofs0[i]].
+    auto apply = [&](auto set_fn)
     {
-      for (std::int32_t idx : _dofs0)
+      std::int32_t x_size = x.size();
+      for (std::size_t i = 0; i < _dofs0.size(); ++i)
       {
-        if (idx < x_size)
-          x[idx] = 0;
+        if (_dofs0[i] < x_size)
+          x[_dofs0[i]] = set_fn(i);
       }
+    };
+
+    if (alpha == T(0)) // Optimisation for when alpha == 0
+    {
+      apply([](std::size_t) -> T { return 0; });
+      return;
     }
-    else
+
+    auto handle_function = [&](std::shared_ptr<const Function<T, U>> g)
     {
-      if (std::holds_alternative<std::shared_ptr<const Function<T, U>>>(_g))
+      assert(g);
+      auto dofs1_g = _dofs1_g.empty() ? std::span(_dofs0) : std::span(_dofs1_g);
```

Comment:
Not necessarily a comment on this pull request. But some documentation here would be nice.

I.e. Why do we need to call the internal variable (to this function) `dofs1_g`, 
when `g` only has 1 set of dofs. I think `dofs_g` is a completely valid name.

I guess the docstring could be something like.

// Extract degrees of freedom associated with `g`. If `g` is in a collapsed sub-space, get the dofs in this space, otherwise the degrees of `g` is the same
as for `_x`